clojure/src/firn/markup.clj

```diff
@@ -2,11 +2,210 @@
+
+(defn render-breadcrumbs
+  "Iterates firn-under, fetching the link to each item, constructuring a breadcrumb
+  in: firn-under: [Research Language English]
+  out             [{:name Research :path /Research} {:name Language :path /Language}]
+  :out-data       [:div [:a {:href path} name] ' > ' [:a {:href path} name]]]] etc."
+  [firn-under sitemap opts]
+  (loop [firn-under   firn-under
+         sitemap-node sitemap
+         out          []]
+    (let [head    (first firn-under)
+          tail    (rest firn-under)
+          from-sm (get sitemap-node head)
+          new-out (conj out {:title head :path (get from-sm :path)})]
+      (if (nil? head)
+        [:div.firn-breadcrumbs
+         ;; construct html
+         (->> out
+            (map #(vec (if (% :path)
+                         [:a {:href (% :path)} (% :title)]
+                         [:span (% :title)])))
+            (interpose [:span (or (opts :separator) " > ")]))]
+        ;; --
+        (recur tail (get-in sitemap [head :children]) new-out)))))
+
+(defn render-adjacent-file
+  "Renders html (or returns data) of the previous and next link in the sitemap.
+  Expects that your files are using `firn-order` or `:date-created-ts` in order to
+  properly determine what is `next` and what is `previous`."
+  [{:keys [sitemap firn-under firn-order date-created-ts as-data? order-by prev-text next-text]
+    :or   {order-by :firn-order}}]
+  (let [site-map-node (if (nil? firn-under) sitemap
+                          (get-in sitemap (u/interpose+tail firn-under :children)))
+        sort-mech     {:date       {:key :date-created-ts :file-val date-created-ts}
+                       :firn-order {:key :firn-order :file-val firn-order}}
+        sort-key      (-> sort-mech order-by :key)
+        org-file-val  (-> sort-mech order-by :file-val)
+        ordered-smn   (->> site-map-node vals (sort-by sort-key))
+        prev-text     (or prev-text "Previous: ")
+        next-text     (or next-text "Next: ")
+        out           (atom {:next nil :previous nil})]
+    (loop [lst  ordered-smn
+           prev nil]
+      (when (seq lst)
+        (let [head (first lst)]
+          ;; when firn-order equals the item we are iterative over.
+          (if (= (sort-key head) org-file-val)
+            (reset! out {:next (second lst) :previous prev})
+            (recur (rest lst) head)))))
+    (if as-data? @out
+        (let [{:keys [previous next]} @out]
+          [:div.firn-file-navigation
+           (when previous
+             [:span.firn-file-nav-prev
+              prev-text [:a {:href (previous :path)} (previous :title)]])
+           " "
+           (when next
+             [:span.firn-file-nav-next
+              next-text [:a {:href (next :path)} (next :title)]])]))))
+
+;; R: Backlinks -------
+
+(defn render-backlinks
+  [{:keys [site-links file site-url]}]
+  (let [org-path-match-file-url? #(let [x (internal-link-handler (% :path) site-url)]
```

Comment:
@teesloane what are you thoughts on handling internal links to notes in `:ignored-dirs` i.e. private notes?